python/setup.py

```diff
@@ -274,6 +274,77 @@ def download_pickle5(pickle5_dir):
                 wzf.close()
 
 
+def patch_isdir():
+    """
+    Python on Windows is having hard times at telling if a symlink is
+    a directory - it can "guess" wrong at times, which bites when
+    finding packages. Replace with a fixed version which unwraps links first.
+    """
+    orig_isdir = os.path.isdir
+
+    def fixed_isdir(path, debug=False):
+        while os.path.islink(path):
+            try:
+                link = os.readlink(path)
+            except OSError:
+                break
+            path = os.path.abspath(os.path.join(os.path.dirname(path), link))
+        return orig_isdir(path)
+
+    os.path.isdir = fixed_isdir
+
+
+def replace_symlinks_with_junctions():
+    _LINKS = {
+        r"ray\new_dashboard": "../../dashboard",
+        r"ray\rllib": "../../rllib",
+        r"ray\streaming": "../../streaming/python/",
+    }
+    root_dir = os.path.dirname(__file__)
+    for link, default in _LINKS.items():
+        path = os.path.join(root_dir, link)
+        try:
+            out = subprocess.check_output(
+                "DIR /A:LD /B", shell=True, cwd=os.path.dirname(path))
+        except subprocess.CalledProcessError:
+            out = b""
+        if os.path.basename(path) in out.decode("utf8").splitlines():
+            logger.info(f"'{link}' is already converted to junction point")
+        else:
+            logger.info(f"Converting '{link}' to junction point...")
+            if os.path.isfile(path):
+                with open(path) as inp:
+                    target = inp.read()
+                os.unlink(path)
+            elif os.path.isdir(path):
+                target = default
+                try:
+                    # unlink() works on links as well as on regular files,
+                    # and links to directories are considered directories now
+                    os.unlink(path)
+                except OSError as err:
+                    # On Windows attempt to unlink a regular directory results
+                    # in a PermissionError with errno set to errno.EACCES.
+                    if err.errno != errno.EACCES:
+                        raise
+                    # For regular directories deletion is done with rmdir call.
+                    os.rmdir(path)
+            else:
+                raise ValueError(f"Unexpected type of entry: '{path}'")
+            target = os.path.abspath(
+                os.path.join(os.path.dirname(path), target))
+            logger.info("Setting {} -> {}".format(link, target))
+            subprocess.check_call(
+                f"MKLINK /J '{os.path.basename(link)}' '{target}'",
+                shell=True,
+                cwd=os.path.dirname(path))
+
+
+if is_native_windows_or_msys():
+    patch_isdir()
+    replace_symlinks_with_junctions()
```

Comment:
I feel that this function should be called under some condition only, because it alters the source tree which might be unexpected for a user. I propose to add either an environment variable or a command line switch to turn it on.

Reply:
Added